cpp/include/cudf/detail/hash_reduce_by_row.cuh

```diff
@@ -14,61 +14,56 @@
  * limitations under the License.
  */
 
-#include "distinct_reduce.cuh"
+#include <stream_compaction/stream_compaction_common.cuh>
```

Comment:
Including a `src` header in `include` is fishy. Can we move the relevant code to this file and get rid of this inclusion?

Reply:
The code in this file needs to use `compaction_hash` in `<src/>` header (`stream_compaction_common.cuh`). I overlooked that. So we must put this `hash_reduce_by_row.cuh` in the `src` folder.

Reply:
`compaction_hash` is a hash value adaptor that avoids getting the sentinel value as the hashing output. This is a generic functionality (not stream compaction specific) and can be placed in the detail header. The `compaction_hash` logic can be removed once we use e.g. `-1` as the key sentinel but we can defer this change when migrating `distinct` to use new cuco set/map.

To make this right and facilitate the future refactoring, I would suggest that we:

- Move the `reduce_by_row.cuh` back to `include`
- Move `compaction_hash` functor to the `reduce_by_row.cuh` file and rename it as `sentinel_hash_adaptor` (or simply `hash_adaptor``)

Reply:
https://github.com/rapidsai/cudf/blob/07845438855c3787dc5daec0195bd75a399cd598/cpp/src/stream_compaction/stream_compaction_common.hpp#L33

If you are worrying about the stream compaction key sentinel value, Just change it to `-1` and we don't need  `compaction_hash` at all.

Reply:
Then there will be some small changes to other files. But that should be good, since we are only removing code.